File: dazzler/system/transforms.py
"""Ties & transform API"""
import typing
from ._binding import Target, coerce_binding


class Transform:
    """
    **Execution order: Recursive Left to Right**
    """

    def __init__(self):
        self.args = {}
        self.next = []

    def transform(self, transform):
        """
        Add a transform to chain after this one and all the previously added
        transforms.

        :param transform: Transformation to apply.
        :type transform: Transform
        :return: self
        """
        self.next.append(transform)
        return self

    def t(self, transform):
        return self.transform(transform)

    def prepare(self):
        return {
            'transform': self.__class__.__name__,
            'args': {
                k: v.prepare() if hasattr(v, 'prepare') else v
                for k, v in self.args.items()
            },
            'next': [n.prepare() for n in self.next]
        }


class TiedTransform:
    def __init__(self, trigger, targets):
        """
        :param trigger: The aspect starting a tie chain on update.
        :type trigger: dazzler.system.Trigger
        :param targets: Aspects to update.
        :type targets: typing.List[dazzler.system.Target]
        """
        self.trigger = trigger
        self.targets = targets
        self.transforms = []

    def transform(self, transform: Transform):
        self.transforms.append(transform)
        return self

    def t(self, transform: Transform):
        return self.transform(transform)

    def prepare(self):
        return {
            'trigger': self.trigger.prepare(),
            'targets': [t.prepare() for t in self.targets],
            'transforms': [t.prepare() for t in self.transforms],
        }


class _Transformable(Transform):
    def __init__(self, transform: Transform):
        super().__init__()
        self.args['transform'] = transform


class _Comparable(Transform):
    def __init__(self, comparison: Transform):
        super().__init__()
        self.args['comparison'] = comparison


class _ValueTransform(Transform):
    def __init__(self, value: typing.Any):
        super().__init__()
        self.args['value'] = value


class _TargetTransform(Transform):
    def __init__(self, target: typing.Union[str, Target]):
        super().__init__()
        self.args['target'] = coerce_binding(target, Target)


class _Comparison(Transform):
    def __init__(self, other):
        super().__init__()
        self.args['other'] = other


# Number transform
class Add(_ValueTransform):
    """Perform an addition on the chain value."""


class Sub(_ValueTransform):
    """Subtract values."""


class Divide(_ValueTransform):
    """Division operation Left / Right"""


class Multiply(_ValueTransform):
    """Multiply chain value * value"""


class Modulus(_ValueTransform):
    """Perform a modulus operation on the chain."""


class ToPrecision(Transform):

    def __init__(self, precision: int):
        super().__init__()
        self.args['precision'] = precision


# String transform

class Format(Transform):
    """
    Format the value into the template.

    Template format: ``${key}``

    For primitive values (string, numbers), the value will be formatted in
    ``${value}`` template key.

    Objects are formatted with their keys items formatted with the value.

    .. code-block:: python

        from dazzler.system.transforms import RawValue, Format

        RawValue('foo').Format('${value} bar')
        # => 'foo bar'

        RawValue({'a': 'foo', 'b': 'bar'}).Format('${a} ${b}')
        # => 'foo bar'

        RawValue(['foo', 'bar']).Format('${0} ${1}')
        # => 'foo bar'
    """

    def __init__(self, template: str):
        super().__init__()
        self.args['template'] = template


class Split(Transform):
    """Split a string value into a list."""
    def __init__(self, separator: str):
        super().__init__()
        self.args['separator'] = separator


class ToLower(Transform):
    """Transform a string chained value to lower case."""


class ToUpper(Transform):
    """Transform a string chained value to upper case."""


class Trim(Transform):
    """Trim whitespace around a chained string value."""


# List transforms

class Concat(Transform):
    """
    Concat a list or string into a new string/list containing entries
    from both sides.
    """

    def __init__(self, other: typing.Union[list, str]):
        super().__init__()
        self.args['other'] = other


class Slice(Transform):
    """Take a slice of a list from ``start`` to ``stop``."""

    def __init__(self, start: int, stop: int):
        super().__init__()
        self.args['start'] = start
        self.args['stop'] = stop


class Map(_Transformable):
    """
    Map a chained array values.
    """


class Reduce(_Transformable):
    """Reduce the list value"""


class Filter(_Comparable):
    """Filter elements of the list"""


class Pluck(Transform):
    """
    Take a value at ``field`` from a list of objects.

    .. code-block:: python

        from dazzler.system.transforms import RawValue, Pluck

        RawValue([{'a': 1, 'b': 2, 'c': 3}]).t(Pick('a'))
        # => [1]
    """
    def __init__(self, field: str):
        super().__init__()
        self.args['field'] = field


class Append(_ValueTransform):
    """Append a value a the end of the chained list."""


class Prepend(_ValueTransform):
    """Prepend a value at the beginning of the list."""


class Insert(_TargetTransform):
    """Insert the value into the target list."""
    def __init__(self, target, front=False):
        super().__init__(target)
        self.args['front'] = front


class Take(Transform):
    """Take the first `n` elements of the list."""
    def __init__(self, n: int):
        super().__init__()
        self.args['n'] = n


class Length(Transform):
    """Resolve the length of the list."""


class Range(Transform):
    """Return an array of number from ``start`` to ``end``"""
    def __init__(self, start: int, end: int, step: int = 1):
        super().__init__()
        self.args['start'] = start
        self.args['end'] = end
        self.args['step'] = step


class Includes(_ValueTransform):
    """True if value is included in the chained list."""


class Find(_Comparable):
    """
    Find an item in a list with a comparison.

    .. code-block:: python

        from dazzler.system.transforms import RawValue, Find, Get, Equals

        RawValue([{'a': 'a', 'b': 'b'}, {'a': 'b', 'b': 'a'}])\
            .transform(Find(Get('a').t(Equals('b'))))
        # => {'a': 'b', 'b': 'a'}
    """


class Join(Transform):
    def __init__(self, separator):
        super().__init__()
        self.args['separator'] = separator


class Reverse(Transform):
    """Reverse a string or array elements."""


class Unique(Transform):
    """Filter duplicates out of the list."""


class Zip(_ValueTransform):
    """Associates values from each list"""


class Sort(_Transformable):
    """Sort transform should return a number"""


class ToPairs(Transform):
    """Transform an object into pairs of [key, value]"""


class FromPairs(Transform):
    """Transform a list of list of key value pairs into an object."""


# Object transforms
class Pick(Transform):
    """
    Pick the fields of the object.

    .. code-block:: python

        from dazzler.system.transforms import RawValue, Pick

        RawValue({'a': 1, 'b': 2, 'c': 3}).t(Pick(['a', 'b']))
        # => {'a': 1, 'b': 2}
    """
    def __init__(self, fields: typing.List[str]):
        super().__init__()
        self.args['fields'] = fields


class Get(Transform):
    """Get the field value of the chain object."""
    def __init__(self, field: str):
        super().__init__()
        self.args['field'] = field


class Set(Transform):
    """
    Set the key value on the trigger value.

    Value can be raw or a :py:class:`dazzler.system.Target` aspect.
    """
    def __init__(self, key: str, value: typing.Any):
        """

        :param key:
        :param value:
        """
        super().__init__()
        self.args['key'] = key
        self.args['value'] = value


class Put(Transform):
    """
    Put the value at key on target

    Target should be a dict or a `~.dazzler.system.Target` aspect
    resolving to a dict.
    """
    def __init__(self, key: str, target):
        super().__init__()
        self.args['key'] = key
        self.args['target'] = target


class Merge(Transform):
    """
    Merge a chained ``dict`` value with another ``dict``, either raw value
    or a :py:class:`~.dazzler.system.Target` aspect.

    .. code-block:: python

        from dazzler.system.transforms import RawValue, Merge

        RawValue({'a': 1}).transform(Merge({'b': 2}))
        # => {'a': 1, 'b': 2}
    """
    def __init__(self, other, direction='right', deep=False):
        super().__init__()
        self.args['other'] = other
        self.args['direction'] = direction
        self.args['deep'] = deep


class ToJson(Transform):
    """Serialize the chain value to JSON."""


class FromJson(Transform):
    """Parse the chain value from JSON."""


# Conditionals

class If(_Comparable):
    """
    Initiate a comparison on the value.

    ``then`` is execute if the comparison is ``True`` else ``otherwise`` will
    be executed if defined.

    .. code-block:: python

        from dazzler.system.transforms import If, RawValue, Equals, Format

        RawValue('foo').t(If(Equals('foo'), then=Format('${value} bar')))
    """
    def __init__(self, comparison, then, otherwise=None):
        super().__init__(comparison)
        self.args['then'] = then
        self.args['otherwise'] = otherwise


class Equals(_Comparison):
    pass


class NotEquals(_Comparison):
    pass


class Match(_Comparison):
    pass


class Greater(_Comparison):
    pass


class GreaterOrEquals(_Comparison):
    pass


class Lesser(_Comparison):
    pass


class LesserOrEquals(_Comparison):
    pass


class And(_Comparison):
    pass


class Or(_Comparison):
    pass


class Not(Transform):
    pass


class RawValue(_ValueTransform):
    """Resolve a raw value."""


class AspectValue(_TargetTransform):
    """Resolve the target value."""

File: src/renderer/js/transforms.ts
/* eslint-disable no-use-before-define */
import {
    concat,
    drop,
    equals,
    find,
    fromPairs,
    includes,
    is,
    join,
    mergeDeepLeft,
    mergeDeepRight,
    mergeLeft,
    mergeRight,
    pick,
    pluck,
    reduce,
    replace,
    reverse,
    slice,
    sort,
    split,
    take,
    toPairs,
    trim,
    uniq,
    zip,
} from 'ramda';
import {Transform, TransformFunc, TransformGetAspectFunc} from './types';
import {coerceAspect, isAspect} from './aspects';

const transforms: {[key: string]: TransformFunc} = {
    /* String transforms */
    ToUpper: (value) => {
        return value.toUpperCase();
    },
    ToLower: (value) => {
        return value.toLowerCase();
    },
    Format: (value, args) => {
        const {template} = args;
        if (is(String, value) || is(Number, value) || is(Boolean, value)) {
            return replace('${value}', value, template);
        } else if (is(Object, value)) {
            return reduce(
                (acc, [k, v]) => replace(`$\{${k}}`, v, acc),
                template,
                toPairs(value)
            );
        }
        return value;
    },
    Split: (value, args) => {
        const {separator} = args;
        return split(separator, value);
    },
    Trim: (value) => {
        return trim(value);
    },
    /* Number Transform */
    Add: (value, args, getAspect) => {
        if (is(Number, args.value)) {
            return value + args.value;
        }
        return value + coerceAspect(args.value, getAspect);
    },
    Sub: (value, args, getAspect) => {
        if (is(Number, args.value)) {
            return value - args.value;
        }
        return value - coerceAspect(args.value, getAspect);
    },
    Divide: (value, args, getAspect) => {
        if (is(Number, args.value)) {
            return value / args.value;
        }
        return value / coerceAspect(args.value, getAspect);
    },
    Multiply: (value, args, getAspect) => {
        if (is(Number, args.value)) {
            return value * args.value;
        }
        return value * coerceAspect(args.value, getAspect);
    },
    Modulus: (value, args, getAspect) => {
        if (is(Number, args.value)) {
            return value % args.value;
        }
        return value % coerceAspect(args.value, getAspect);
    },
    ToPrecision: (value, args) => {
        return value.toPrecision(args.precision);
    },
    /* Array transforms  */
    Concat: (value, args, getAspect) => {
        const {other} = args;
        return concat(value, coerceAspect(other, getAspect));
    },
    Slice: (value, args) => {
        return slice(args.start, args.stop, value);
    },
    Map: (value, args, getAspect) => {
        const {transform} = args;
        return value.map((e) =>
            executeTransform(
                transform.transform,
                e,
                transform.args,
                transform.next,
                getAspect
            )
        );
    },
    Filter: (value, args, getAspect) => {
        const {comparison} = args;
        return value.filter((e) =>
            executeTransform(
                comparison.transform,
                e,
                comparison.args,
                comparison.next,
                getAspect
            )
        );
    },
    Reduce: (value, args, getAspect) => {
        const {transform, accumulator} = args;
        const acc = coerceAspect(accumulator, getAspect);
        return value.reduce(
            (previous, next) =>
                executeTransform(
                    transform.transform,
                    [previous, next],
                    transform.args,
                    transform.next,
                    getAspect
                ),
            acc
        );
    },
    Pluck: (value, args) => {
        const {field} = args;
        return pluck(field, value);
    },
    Append: (value, args, getAspect) => {
        return concat(value, [coerceAspect(args.value, getAspect)]);
    },
    Prepend: (value, args, getAspect) => {
        return concat([coerceAspect(args.value, getAspect)], value);
    },
    Insert: (value, args, getAspect) => {
        const {target, front} = args;
        const t = coerceAspect(target, getAspect);
        return front ? concat([value], t) : concat(t, [value]);
    },
    Take: (value, args, getAspect) => {
        const {n} = args;
        return take(coerceAspect(n, getAspect), value);
    },
    Length: (value) => {
        return value.length;
    },
    Range: (value, args, getAspect) => {
        const {start, end, step} = args;
        const s = coerceAspect(start, getAspect);
        const e = coerceAspect(end, getAspect);
        let i = s;
        const arr = [];
        while (i < e) {
            arr.push(i);
            i += step;
        }
        return arr;
    },
    Includes: (value, args, getAspect) => {
        return includes(coerceAspect(args.value, getAspect), value);
    },
    Find: (value, args, getAspect) => {
        const {comparison} = args;
        return find((a) =>
            executeTransform(
                comparison.transform,
                a,
                comparison.args,
                comparison.next,
                getAspect
            )
        )(value);
    },
    Join: (value, args, getAspect) => {
        return join(coerceAspect(args.separator, getAspect), value);
    },
    Sort: (value, args, getAspect) => {
        const {transform} = args;
        return sort(
            (a, b) =>
                executeTransform(
                    transform.transform,
                    [a, b],
                    transform.args,
                    transform.next,
                    getAspect
                ),
            value
        );
    },
    Reverse: (value) => {
        return reverse(value);
    },
    Unique: (value) => {
        return uniq(value);
    },
    Zip: (value, args, getAspect) => {
        return zip(value, coerceAspect(args.value, getAspect));
    },
    /* Object transforms */
    Pick: (value, args) => {
        return pick(args.fields, value);
    },
    Get: (value, args) => {
        return value[args.field];
    },
    Set: (v, args, getAspect) => {
        const {key, value} = args;
        v[key] = coerceAspect(value, getAspect);
        return v;
    },
    Put: (value, args, getAspect) => {
        const {key, target} = args;
        const obj = coerceAspect(target, getAspect);
        obj[key] = value;
        return obj;
    },
    Merge: (value, args, getAspect) => {
        const {deep, direction, other} = args;
        let otherValue = other;
        if (isAspect(other)) {
            otherValue = getAspect(other.identity, other.aspect);
        }
        if (direction === 'right') {
            if (deep) {
                return mergeDeepRight(value, otherValue);
            }
            return mergeRight(value, otherValue);
        }
        if (deep) {
            return mergeDeepLeft(value, otherValue);
        }
        return mergeLeft(value, otherValue);
    },
    ToJson: (value) => {
        return JSON.stringify(value);
    },
    FromJson: (value) => {
        return JSON.parse(value);
    },
    ToPairs: (value) => {
        return toPairs(value);
    },
    FromPairs: (value) => {
        return fromPairs(value);
    },
    /* Conditionals */
    If: (value, args, getAspect) => {
        const {comparison, then, otherwise} = args;
        const c = transforms[comparison.transform];

        if (c(value, comparison.args, getAspect)) {
            return executeTransform(
                then.transform,
                value,
                then.args,
                then.next,
                getAspect
            );
        }
        if (otherwise) {
            return executeTransform(
                otherwise.transform,
                value,
                otherwise.args,
                otherwise.next,
                getAspect
            );
        }
        return value;
    },
    Equals: (value, args, getAspect) => {
        return equals(value, coerceAspect(args.other, getAspect));
    },
    NotEquals: (value, args, getAspect) => {
        return !equals(value, coerceAspect(args.other, getAspect));
    },
    Match: (value, args, getAspect) => {
        const r = new RegExp(coerceAspect(args.other, getAspect));
        return r.test(value);
    },
    Greater: (value, args, getAspect) => {
        return value > coerceAspect(args.other, getAspect);
    },
    GreaterOrEquals: (value, args, getAspect) => {
        return value >= coerceAspect(args.other, getAspect);
    },
    Lesser: (value, args, getAspect) => {
        return value < coerceAspect(args.other, getAspect);
    },
    LesserOrEquals: (value, args, getAspect) => {
        return value <= coerceAspect(args.other, getAspect);
    },
    And: (value, args, getAspect) => {
        return value && coerceAspect(args.other, getAspect);
    },
    Or: (value, args, getAspect) => {
        return value || coerceAspect(args.other, getAspect);
    },
    Not: (value) => {
        return !value;
    },
    RawValue: (value, args) => {
        return args.value;
    },
    AspectValue: (value, args, getAspect) => {
        const {identity, aspect} = args.target;
        return getAspect(identity, aspect);
    },
};

export const executeTransform = (
    transform: string,
    value: any,
    args: any,
    next: Array<Transform>,
    getAspect: TransformGetAspectFunc
) => {
    const t = transforms[transform];
    const newValue = t(value, args, getAspect);
    if (next.length) {
        const n = next[0];
        return executeTransform(
            n.transform,
            newValue,
            n.args,
            // Execute the next first, then back to chain.
            concat(n.next, drop(1, next)),
            getAspect
        );
    }
    return newValue;
};

export default transforms;
